File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/telemetry/Telemetry.java
package org.firstinspires.ftc.teamcode.telemetry;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * TODO Document
 */

public class Telemetry implements ITelemetry {
    private org.firstinspires.ftc.robotcore.external.Telemetry telem;
    private LinkedHashMap<String, Object> buffer = new LinkedHashMap<>();

    public Telemetry(org.firstinspires.ftc.robotcore.external.Telemetry telem) {
        this.telem = telem;
        this.telem.addData("Hello World", "Telemetry Initialized!");
        this.telem.update();
    }

    @Override
    public void write(String caption, String data) {
        this.buffer.put(caption, data);
        wrapUp();
    }

    @Override
    public void error(String info) {
        this.buffer.put("[ERROR]", info);
        wrapUp();
    }

    @Override
    public void warning(String info) {
        this.buffer.put("[WARN]", info);
        wrapUp();
    }

    @Override
    public void data(String label, Object data) {
        this.buffer.put("DATA: " + label, data);
        wrapUp();
    }

    @Override
    public void fatal(String info) {
        this.buffer.put("--- FATAL ", " ERROR ---");
        this.buffer.put("Error Info", info);
        wrapUp();
    }

    @Override
    public void clear() {
        this.buffer.clear();
        wrapUp();
    }

    private void wrapUp() {
        for (Map.Entry<String, Object> entry : this.buffer.entrySet()) {
            telem.addData(entry.getKey(), entry.getValue());
        }
        telem.update();
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/telemetry/Telemetry.kt
package org.firstinspires.ftc.teamcode.telemetry

/**
 * A reference implementation of ITelemetry.
 * @author Michael Peng
 */
class Telemetry(private val telem: org.firstinspires.ftc.robotcore.external.Telemetry) : ITelemetry {

    override var autoClear: Boolean
        get() = telem.isAutoClear
        set(value) {
            telem.isAutoClear = value
        }

    override var autoUpdate = false

    init {
        this.telem.addData("Hello World", "Telemetry Initialized!")
        flush()
    }

    override fun write(caption: String, data: String) {
        this.telem.addData(caption, data)
        if (autoUpdate) flush()
    }

    override fun error(info: String) {
        this.telem.addData("[ERROR]", info)
        if (autoUpdate) flush()
    }

    override fun warning(info: String) {
        this.telem.addData("[WARN]", info)
        if (autoUpdate) flush()
    }

    override fun data(label: String, data: Any) {
        this.telem.addData("DATA: " + label, data)
        if (autoUpdate) flush()
    }

    override fun fatal(info: String) {
        this.telem.addData("--- FATAL ", " ERROR ---")
        this.telem.addData("Error Info", info)
        if (autoUpdate) flush()
    }

    // FIXME("repetition") Is there a way to prevent repetition of any kind that accomplishes autoUpdate logic?
    //   Failed solutions:
    //     if (autoUpdate) flush()
    //     fun maybeUpdate() = if (autoUpdate) flush()

    override fun flush() {
        this.telem.update()
    }
}